File: scripts/verify-market-category.ts
/**
 * verify:market-category — proof harness for the observed-trade-category-field
 * card (approved 2026-09-23). Two claims, both empirical:
 *
 *   1. PARITY: the TypeScript classifier (src/lib/market-category.ts) labels the
 *      same rows exactly as the validated Python reference implementation
 *      (scripts/wallet-concentration-test.py::classify). Mismatches are printed
 *      and fail the run.
 *   2. WHAT THE FIELD MEANT: over the same stratified sample, how much of the
 *      stored `marketCategory` token is a question word rather than a market
 *      family, and what the market actually is.
 *
 * Usage:
 *   python3 scripts/verify-market-category.py        # reference labels
 *   npx tsx scripts/verify-market-category.ts        # diff + census
 */

import { classifyMarketCategory } from "../src/lib/market-category";
import * as fs from "fs";
import { join } from "path";

const ROOT = join(__dirname, "..");
const REF = join(ROOT, "data", "market-category-reference.json");
const OUT = join(ROOT, "data", "market-category-verification.json");

const QUESTION_WORDS = new Set([
  "will", "would", "what", "which", "who", "when", "where", "how", "why", "is", "are", "was",
  "were", "can", "could", "does", "do", "did", "at", "by", "on", "in", "of", "the", "to", "for",
  "highest", "lowest", "most", "least", "best", "worst", "next", "first", "last", "new", "any",
  "all", "this", "that", "if", "before", "after", "over", "under", "between", "during", "than",
  "more", "less", "top", "no", "not", "yes", "us", "up", "down", "who-will", "will-the",
]);

function isQuestionWord(token: string | null): boolean {
  if (!token) return false;
  const parts = token.toLowerCase().split(/[-_\s]+/).filter(Boolean);
  return parts.some((p) => QUESTION_WORDS.has(p));
}

type RefRow = {
  marketId: string;
  question: string;
  token: string | null;
  refCoarse: string;
  refFine: string;
};

function main() {
  if (!fs.existsSync(REF)) {
    throw new Error(`Missing ${REF} — run: python3 scripts/verify-market-category.py`);
  }
  const { rows, generatedAt, classifier } = JSON.parse(fs.readFileSync(REF, "utf-8")) as {
    rows: RefRow[];
    generatedAt: string;
    classifier: string;
  };

  const mismatches: Array<{ marketId: string; expected: string; got: string }> = [];
  const coarse = new Map<string, number>();
  const fine = new Map<string, number>();
  const tokenHit = new Map<string, number>();
  let questionWordRows = 0;
  let tokenAlwaysOther = 0;
  let tokenSpansManyClasses = 0;
  const tokenClasses = new Map<string, Set<string>>();

  for (const r of rows) {
    const got = classifyMarketCategory(r.marketId, r.question);
    if (got.coarse !== r.refCoarse || got.fine !== r.refFine) {
      mismatches.push({ marketId: r.marketId, expected: `${r.refCoarse}/${r.refFine}`, got: `${got.coarse}/${got.fine}` });
    }
    coarse.set(got.coarse, (coarse.get(got.coarse) ?? 0) + 1);
    fine.set(got.fine, (fine.get(got.fine) ?? 0) + 1);
    const tok = r.token ?? "(null)";
    tokenHit.set(tok, (tokenHit.get(tok) ?? 0) + 1);
    if (isQuestionWord(r.token)) questionWordRows++;
    if (!tokenClasses.has(tok)) tokenClasses.set(tok, new Set());
    tokenClasses.get(tok)!.add(got.coarse);
  }

  for (const tok of tokenHit.keys()) {
    const classes = tokenClasses.get(tok)!;
    if (classes.size === 1 && classes.has("other")) tokenAlwaysOther++;
    if (classes.size > 2) tokenSpansManyClasses++;
  }

  const total = rows.length;
  const pc = (n: number) => `${((100 * n) / total).toFixed(1)}%`;
  const topEntries = (m: Map<string, number>, n = 12) =>
    [...m.entries()].sort((a, b) => b[1] - a[1]).slice(0, n);

  console.log(`\nSample: ${total} stratified rows | reference classifier: ${classifier} (${generatedAt})`);
  console.log(`\nPARITY: ${mismatches.length === 0 ? "PASS — 0 mismatches" : `FAIL — ${mismatches.length} mismatch(es)`}`);
  for (const m of mismatches.slice(0, 10)) console.log(`   ${m.marketId.slice(0, 50)} expected ${m.expected} got ${m.got}`);

  console.log(`\nWHAT THE STORED TOKEN IS`);
  console.log(`  rows whose token is a question/function word: ${questionWordRows} (${pc(questionWordRows)})`);
  console.log(`  tokens that classify to 'other' only: ${tokenAlwaysOther} of ${tokenHit.size} distinct tokens`);
  console.log(`  tokens spanning >2 real categories (i.e. not a category at all): ${tokenSpansManyClasses}`);
  console.log(`  top tokens in sample: ${topEntries(tokenHit, 8).map(([t, n]) => `${t}=${n}`).join(", ")}`);

  console.log(`\nWHAT THE MARKET ACTUALLY IS (coarse category over the same rows)`);
  for (const [c, n] of topEntries(coarse, 12)) console.log(`  ${c.padEnd(10)} ${String(n).padStart(6)}  ${pc(n)}`);
  console.log(`\nfine grain, top 12:`);
  for (const [c, n] of topEntries(fine, 12)) console.log(`  ${c.padEnd(18)} ${String(n).padStart(6)}  ${pc(n)}`);

  fs.writeFileSync(
    OUT,
    JSON.stringify(
      {
        generatedAt: new Date().toISOString(),
        reference: { classifier, generatedAt, rows: total },
        parity: { passed: mismatches.length === 0, mismatchCount: mismatches.length, mismatches: mismatches.slice(0, 50) },
        tokenSemantics: {
          questionWordRows,
          questionWordShare: questionWordRows / total,
          tokensClassifyingToOtherOnly: tokenAlwaysOther,
          tokensSpanningMoreThanTwoCategories: tokenSpansManyClasses,
          distinctTokens: tokenHit.size,
        },
        classes: { coarse: Object.fromEntries(topEntries(coarse, 50)), fine: Object.fromEntries(topEntries(fine, 50)) },
      },
      null,
      2
    )
  );
  console.log(`\nReport → ${OUT}`);
  if (mismatches.length > 0) process.exitCode = 1;
}

main();

File: scripts/verify-market-category.py
#!/usr/bin/env python3
"""verify-market-category (reference side) — observed-trade-category-field card.

Emits a STRATIFIED sample of stored observed trades with the label the validated
Python classifier gives them (scripts/wallet-concentration-test.py::classify, the
implementation that agreed with the vendor's own category map at r = +0.548 over
353 wallets). The TypeScript port is then diffed against this file by
scripts/verify-market-category.ts — parity is verified, not assumed.

  python3 scripts/verify-market-category.py            # -> data/market-category-reference.json

Read-only against prisma/dev.db.
"""
import importlib.util
import json
import os
import random
import sqlite3
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
# mode=ro, deliberately NOT immutable=1: the DB now runs journal_mode=wal (tuning #35
# Rec 1, applied 2026-09-24), and immutable=1 tells SQLite the file can never change, so
# it would skip the WAL entirely and read a stale (possibly pre-checkpoint) snapshot.
# Read-only + busy_timeout reads the current committed state without taking write locks.
DB = f"file:{os.path.join(ROOT, 'prisma', 'dev.db')}?mode=ro"
OUT = os.path.join(ROOT, "data", "market-category-reference.json")
PER_TOKEN = 200
RANDOM_ROWS = 2000
SEED = 20260923


def load_classifier():
    path = os.path.join(ROOT, "scripts", "wallet-concentration-test.py")
    spec = importlib.util.spec_from_file_location("conc_instrument", path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod.classify


def main():
    classify = load_classifier()
    db = sqlite3.connect(DB, uri=True)
    c = db.cursor()
    rows = {}
    top = [r[0] for r in c.execute(
        "SELECT marketCategory FROM ObservedTrade WHERE marketCategory IS NOT NULL "
        "GROUP BY 1 ORDER BY COUNT(*) DESC LIMIT 30"
    )]
    for tok in top:
        for mid, q in c.execute(
            "SELECT marketId, marketQuestion FROM ObservedTrade WHERE marketCategory=? LIMIT ?", (tok, PER_TOKEN)
        ):
            rows[(mid, q)] = tok
    rnd = random.Random(SEED)
    total = c.execute("SELECT COUNT(*) FROM ObservedTrade").fetchone()[0]
    for _ in range(RANDOM_ROWS):
        off = rnd.randrange(total)
        r = c.execute(
            "SELECT marketId, marketQuestion, marketCategory FROM ObservedTrade LIMIT 1 OFFSET ?", (off,)
        ).fetchone()
        rows[(r[0], r[1])] = r[2]

    out = []
    for (mid, q), tok in rows.items():
        coarse, fine = classify(mid, q)
        out.append({
            "marketId": mid,
            "question": q,
            "token": tok,
            "refCoarse": coarse,
            "refFine": fine,
        })
    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    with open(OUT, "w") as f:
        json.dump({"generatedAt": __import__("datetime").datetime.now().isoformat(),
                   "classifier": "scripts/wallet-concentration-test.py::classify",
                   "rows": out}, f)
    print(f"wrote {OUT}: {len(out)} stratified rows ({len(top)} tokens x {PER_TOKEN} + {RANDOM_ROWS} random)")


if __name__ == "__main__":
    sys.exit(main())
